contracts/src/core/libraries/v3/PaymentVaultLib.sol

```diff
@@ -0,0 +1,199 @@
+    function checkReservation(
+        uint64 quorumId,
+        address account,
+        PaymentVaultTypes.Reservation memory reservation,
+        uint64 schedulePeriod
+    ) internal view returns (uint64 startTimestamp) {
+        PaymentVaultTypes.Quorum storage quorum = s().quorum[quorumId];
+        PaymentVaultTypes.User storage user = quorum.user[account];
+        require(reservation.startTimestamp % schedulePeriod == 0, "Invalid start timestamp");
+        require(reservation.endTimestamp % schedulePeriod == 0, "Invalid end timestamp");
+        require(reservation.endTimestamp > reservation.startTimestamp, "Invalid reservation period");
+        require(
+            reservation.endTimestamp - reservation.startTimestamp <= quorum.protocolCfg.reservationAdvanceWindow,
+            "Reservation period too long"
+        );
+        // If the reservation is not expired, the reservation can only be updated favorably to the user.
+        if (block.timestamp <= reservation.endTimestamp) {
+            PaymentVaultTypes.Reservation memory currentReservation = user.reservation;
+            require(
+                reservation.startTimestamp == currentReservation.startTimestamp
+                    && reservation.endTimestamp > currentReservation.endTimestamp,
+                "Invalid reservation update"
+            );
+            require(reservation.symbolsPerSecond >= currentReservation.symbolsPerSecond, "Invalid symbols per second");
+            return currentReservation.endTimestamp;
+        }
+        return reservation.startTimestamp;
+    }
+
+    function depositOnDemand(uint64 quorumId, address account, uint256 amount) internal {
+        PaymentVaultTypes.Quorum storage quorum = s().quorum[quorumId];
+        PaymentVaultTypes.User storage user = quorum.user[account];
+        PaymentVaultTypes.QuorumConfig storage cfg = quorum.cfg;
+
+        uint256 newAmount = user.deposit + amount;
+        require(newAmount <= type(uint80).max, "Amount too large");
+
+        IERC20(cfg.token).safeTransferFrom(account, cfg.recipient, amount);
+
+        user.deposit = newAmount;
+    }
+
+    function populateReservedSymbols(
+        uint64 quorumId,
+        uint64 startTimestamp,
+        uint64 endTimestamp,
+        uint64 symbolsPerSecond,
+        uint64 schedulePeriod
+    ) internal {
+        require(startTimestamp % schedulePeriod == 0, "Invalid start timestamp");
+        require(endTimestamp % schedulePeriod == 0, "Invalid end timestamp");
+        uint64 startPeriod = startTimestamp / schedulePeriod;
+        uint64 endPeriod = endTimestamp / schedulePeriod;
+
+        PaymentVaultTypes.Quorum storage quorum = s().quorum[quorumId];
+        uint64 maxReservedSymbols = quorum.cfg.reservationSymbolsPerSecond;
+        for (uint64 i = startPeriod; i < endPeriod; i++) {
+            uint64 reservedSymbols = quorum.reservedSymbols[i] + symbolsPerSecond;
```

Comment:
if there is an existing reservation that's getting updated, either increase or decrease, I think we will be double counting here? should we subtract the existing rate if there's one?